File: DAL/Repositories/Repository.cs
using DAL.Interfaces;
using DAL.Models;
using DAL.Models.Base;
using DAL.Models.Patient;
using DAL.Models.Test;
using Microsoft.Extensions.Options;
using MongoDB.Driver;

namespace DAL.Repositories;

public class Repository<T> : IRepository<T> where T : Entity, new()
{
    private readonly IMongoCollection<T> _collection;

    protected Repository(IOptions<DbConfig> context, string collectionName)
    {
        var mongoClient = new MongoClient(
            context.Value.ConnectionString);

        var mongoDatabase = mongoClient.GetDatabase(
            context.Value.DatabaseName);

        _collection = mongoDatabase.GetCollection<T>(collectionName);
    }

    public async Task<List<T>> GetAsync() =>
        await _collection.Find(_ => true).ToListAsync();

    public async Task<T?> GetAsync(string id) =>
        await _collection.Find(x => x.Id == id).FirstOrDefaultAsync();

    public async Task<T> CreateAsync(T newBook)
    {
        await _collection.InsertOneAsync(newBook);
        return newBook;
    }

    public async Task UpdateAsync(string id, T updatedBook) =>
        await _collection.ReplaceOneAsync(x => x.Id == id, updatedBook);

    public async Task RemoveAsync(string id) =>
        await _collection.DeleteOneAsync(x => x.Id == id);
}

public class TestsRepository : Repository<Test>
{
    public TestsRepository(IOptions<DbConfig> context) :
        base(context, "tests")
    {
        
    }
}

public class DevicesRepository : Repository<Device>
{
    public DevicesRepository(IOptions<DbConfig> context) :
        base(context, "devices")
    {
        
    }
}

public class CodesRepository : Repository<DeviceLinkInfo>
{
    public CodesRepository(IOptions<DbConfig> context) :
        base(context, "codes")
    {
        
    }
}

public class RegionsRepository : Repository<Region>
{
    public RegionsRepository(IOptions<DbConfig> context) :
        base(context, "regions")
    {
        
    }
}

public class MedicalInstitutionsRepository : Repository<MedicalInstitution>
{
    public MedicalInstitutionsRepository(IOptions<DbConfig> context) :
        base(context, "medical_institutions")
    {
        
    }
}

public class PatientsRepository : Repository<Patient>
{
    public PatientsRepository(IOptions<DbConfig> context) :
        base(context, "patients")
    {
        
    }
}

public class TestStatusesRepository : Repository<TestStatus>
{
    public TestStatusesRepository(IOptions<DbConfig> context) :
        base(context, "test_statuses")
    {
        
    }
}

public class PatientStatusesRepository : Repository<PatientStatus>
{
    public PatientStatusesRepository(IOptions<DbConfig> context) :
        base(context, "patient_statuses")
    {
        
    }
}

public class PatientSexesRepository : Repository<Sex>
{
    public PatientSexesRepository(IOptions<DbConfig> context) :
        base(context, "sexes")
    {
        
    }
}

public class HistoryRepository : Repository<Record>
{
    public HistoryRepository(IOptions<DbConfig> context) :
        base(context, "history")
    {
        
    }
}

File: WEBAPP/ClientApp/src/api/repositories/Repository.ts
import {IRepository} from "../interfaces/IRepository";
import {Region} from "../models/Region";
import {Patient} from "../models/Patient";
import {Device} from "../models/Device";
import {Code} from "../models/Code";
import {PatientStatus} from "../models/PatientStatus";
import {Sex} from "../models/Sex";
import {Record} from "../models/Record";
import {BASE_URL} from "../constants/Constants";
import {MedicalInstitution} from "../models/MedicalInstitution";

abstract class Repository<T> implements IRepository<T> {
    baseUrl: string = BASE_URL;
    collection: string;

    protected constructor(collection: string) {
        this.collection = collection + "/";
    }

    GetAsync(): Promise<T[]> {
        return fetch(this.baseUrl+this.collection)
            .then(r => r.json());
    }
    
    GetItemAsync(id: string): Promise<T> {
        return fetch(this.baseUrl+this.collection+id)
            .then(r => r.json());
    }

    UpdateAsync(id: string, e: T): Promise<void> {
        return fetch(this.baseUrl+this.collection+id, {
            method: "PUT",
            headers: {
                "Content-Type": "application/json"
            },
            body: JSON.stringify(e),
            credentials: "include"
        }).then(_ => {});
    }
    
    DeleteAsync(id: string): Promise<void> {
        return fetch(this.baseUrl+this.collection+id, {
            method: "DELETE",
            headers: {
                "Content-Type": "application/json"
            },
            credentials: "include"
        }).then(_ => {});
    }
}

export class RegionsRepository extends Repository<Region> {
    constructor(collection: string) {
        super(collection);
    }
}

export class PatientsRepository extends Repository<Patient> {
    constructor(collection: string) {
        super(collection);
    }
}

export class PatientStatusesRepository extends Repository<PatientStatus> {
    constructor(collection: string) {
        super(collection);
    }
}

export class PatientSexesRepository extends Repository<Sex> {
    constructor(collection: string) {
        super(collection);
    }
}

class PreLinkInfo {
    region: string
    institution: string
    
    constructor(region: string, institution: string) {
        this.region = region
        this.institution = institution
    }
}

export class DevicesRepository extends Repository<Device> {
    constructor(collection: string) {
        super(collection);
    }

    GetCode(info: PreLinkInfo): Promise<Code> {
        return fetch(this.baseUrl+this.collection+"prelink", {
            method: "POST",
            body: JSON.stringify(info),
            headers: {
                "Content-Type": "application/json"
            }
        })
            .then(r => r.json());
    }
}

export class HistoryRepository extends Repository<Record> {
    constructor(collection: string) {
        super(collection);
    }
    
    GetHistoryForPatient(patientId: string) : Promise<Record[]> {
        return fetch(this.baseUrl+this.collection+`patient/${patientId}`).then(r => r.json())
    }
}

export class MedicalInstitutionsRepository extends Repository<MedicalInstitution> {
    constructor(collection: string) {
        super(collection);
    }
    
    GetAsyncByRegion(regionId: string) : Promise<MedicalInstitution[]> {
        return fetch(this.baseUrl+this.collection+`region/${regionId}`).then(r => r.json())
    }
}
